File: flexinput/src/main/java/com/lytefast/flexinput/managers/SimpleFileManager.java
package com.lytefast.flexinput.managers;

import android.content.Context;
import android.net.Uri;
import android.os.Environment;
import android.support.v4.content.FileProvider;
import android.util.Log;

import java.io.File;
import java.text.SimpleDateFormat;
import java.util.Date;


/**
 * A basic implementation of the {@link FileManager} that uses a specified {@link FileProvider}.
 *
 * @author Sam Shih
 */
@SuppressWarnings("WeakerAccess")
public class SimpleFileManager implements FileManager {

  private static final String TAG = SimpleFileManager.class.getCanonicalName();

  protected final String providerAuthority;
  protected final String imageDirName;
  protected final String imagePrefix;

  public SimpleFileManager(String providerAuthority, String imageDirName) {
    this(providerAuthority, imageDirName, "JPEG_");
  }

  public SimpleFileManager(String providerAuthority, String imageDirName, String imagePrefix) {
    this.providerAuthority = providerAuthority;
    this.imageDirName = imageDirName;
    this.imagePrefix = imagePrefix;
  }

  protected File getImagesDirectory() {
    File file = new File(
        Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES),
        imageDirName);
    if (!file.mkdirs() && !file.isDirectory()) {
      Log.e(TAG, "Directory not created");
    }
    return file;
  }

  @Override
  public File newImageFile() {
    String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss").format(new Date());
    String imageFileName = imagePrefix + timeStamp + ".jpg";
    return new File(getImagesDirectory(), imageFileName);
  }

  @Override
  public Uri toFileProviderUri(final Context context, final File file) {
    return FileProvider.getUriForFile(context, providerAuthority, file);
  }
}

File: flexinput/src/main/java/com/lytefast/flexinput/managers/SimpleFileManager.kt
package com.lytefast.flexinput.managers

import android.content.Context
import android.net.Uri
import android.os.Environment
import android.util.Log
import androidx.core.content.FileProvider
import java.io.File
import java.text.SimpleDateFormat
import java.util.*


/**
 * A basic implementation of the [FileManager] that uses a specified [FileProvider].
 *
 * @author Sam Shih
 */
@Suppress("MemberVisibilityCanBePrivate")
open class SimpleFileManager @JvmOverloads constructor(
    protected val providerAuthority: String,
    protected val imageDirName: String,
    protected val imagePrefix: String = "JPEG_") : FileManager {

  protected val imagesDirectory: File
    get() {
      val file = File(
          Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES),
          imageDirName)
      if (!file.mkdirs() && !file.isDirectory) {
        Log.e(TAG, "Directory not created")
      }
      return file
    }

  override fun newImageFile(): File {
    val timeStamp = SimpleDateFormat("yyyyMMdd_HHmmss").format(Date())
    val imageFileName = "$imagePrefix$timeStamp.jpg"
    return File(imagesDirectory, imageFileName)
  }

  override fun toFileProviderUri(context: Context, file: File): Uri =
      FileProvider.getUriForFile(context, providerAuthority, file)

  companion object {
    private val TAG = SimpleFileManager::class.java.canonicalName
  }
}
